types/react-onclickoutside/index.d.ts

```diff
@@ -1,29 +1,43 @@
 // Type definitions for react-onclickoutside v5.7.0
 // Project: https://github.com/Pomax/react-onclickoutside
-// Definitions by: Karol Janyst <https://github.com/LKay>
+// Definitions by: Karol Janyst <https://github.com/LKay> and cwmoo740 <https://github.com/cwmoo740>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 2.1
 
-import * as React from "react";
+import * as React from 'react';
 
-declare namespace ReactOnClickOutside {
-    interface OnClickOutsideComponent {
-        handleClickOutside(e: React.MouseEvent<any>): void
+declare namespace ClickOut {
+    interface HandleClickOutside<T> {
+        handleClickOutside: React.MouseEventHandler<T>;
     }
 
-    interface OnClickOutsideProps {
-        disableOnClickOutside?: boolean | Function
-        enableOnClickOutside?: Function
-        eventTypes?: string | Array<string>
-        outsideClickIgnoreClass?: string
-        preventDefault?: boolean
-        stopPropagation?: boolean
+    interface InjectedProps {
+        disableOnClickOutside(): void;
+        enableOnClickOutside(): void;
     }
 
-    interface onClickOutside {
-        <A>(component: React.ComponentClass<A> | React.StatelessComponent<A>): React.ComponentClass<A & OnClickOutsideProps>
+    interface Props {
+        disableOnClickOutside: boolean;
+        eventTypes: string | string[];
+        outsideClickIgnoreClass: string;
+        preventDefault: boolean;
+        stopPropagation: boolean;
     }
 }
 
-declare const onClickOutside: ReactOnClickOutside.onClickOutside
-export = onClickOutside;
\ No newline at end of file
+/**
+ * Wraps a React.Stateless or React.Component with clickout
+ * Must pass handleClickOutside as a prop or implement ClickOut.HandleClickOutside
+ * @param component
+ * @constructor
+ */
+declare function ClickOut<OwnProps>(
+    component: Stateless<OwnProps> | Component<OwnProps>
+): React.ComponentClass<OwnProps & Partial<ClickOut.Props>>;
+
+type Stateless<T> = React.StatelessComponent<T & ClickOut.InjectedProps & ClickOut.HandleClickOutside<any>>;
```

Comment:
@cwmoo740 
I came up with this:
```typescript
// Type definitions for react-onclickoutside 5.7
// Project: https://github.com/Pomax/react-onclickoutside
// Definitions by: Karol Janyst <https://github.com/LKay>, cwmoo740 <https://github.com/cwmoo740>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
// TypeScript Version: 2.1

import * as React from "react";
import { Component, ComponentClass, MouseEventHandler, StatelessComponent } from "react";

export = OnClickOut;

declare namespace OnClickOut {
    export interface HandleClickOutside<T> {
        handleClickOutside: MouseEventHandler<T>;
    }

    export interface InjectedOnClickOutProps {
        disableOnClickOutside(): void;
        enableOnClickOutside(): void;
    }

    export interface OnClickOutProps {
        disableOnClickOutside?: boolean;
        eventTypes?: string | string[];
        outsideClickIgnoreClass?: string;
        preventDefault?: boolean;
        stopPropagation?: boolean;
    }
}

/**
 * Wraps a `StatelessComponent` or `ComponentClass` with clickout
 * Must pass handleClickOutside as a prop or implement OnClickOut.HandleClickOutside
 * @param component
 * @constructor
 */

type DecoratedComponent<P> = ComponentClass<P & OnClickOut.OnClickOutProps>;
type WrappedComponentClass<P> =
    //ComponentClass<P & OnClickOut.InjectedOnClickOutProps> |
    ComponentClass<P & OnClickOut.InjectedOnClickOutProps & OnClickOut.HandleClickOutside<any>>;

interface OnClickOut<P> {
    (component: StatelessComponent<P & OnClickOut.InjectedOnClickOutProps & OnClickOut.HandleClickOutside<any>>): DecoratedComponent<P>;
    (component: WrappedComponentClass<P>): DecoratedComponent<P>;
}

declare const OnClickOut: OnClickOut<any>;
```
Couldn't find a good way to check if `handleClickOutside` is define on instance, though. Might be difficult to achieve that with typescript right now.